### PR TITLE
Feature: emit event syscall for Cairo1

### PR DESCRIPTION
**Pull request by maciejka**
Implements Cairo1 `emit event` syscall. Depends on #94 being merged.

### CODE REVIEW

File: src/execution/syscalls.rs [hunk 1]
```diff
@@ -107,34 +108,33 @@ pub fn call_contract(
...
-//     Ok(EmitEventResponse {})
-// }
+
+type EmitEventRequest = EmptyResponse;
+
+impl SyscallRequest for EmitEventRequest {
+    fn read(vm: &VirtualMachine, ptr: &mut Relocatable) -> SyscallResult<EmitEventRequest> {
+        ignore_felt_array::<SyscallExecutionError>(vm, ptr)?;
+        ignore_felt_array::<SyscallExecutionError>(vm, ptr)?;
+        Ok(EmitEventRequest {})
```
Comment (odesenfans) [inline src/execution/syscalls.rs:124]: Where is this syscall defined in `cairo-lang`?

File: src/execution/syscalls.rs [hunk 2]
```diff
@@ -107,34 +108,33 @@ pub fn call_contract(
 //     Ok(DeployResponse { contract_address: deployed_contract_address, constructor_retdata })
 // }
 
-// TODO: EmitEvent syscall.
 // #[derive(Debug, Eq, PartialEq)]
 // pub struct EmitEventRequest {
-//     pub content: EventContent,
+//     pub keys: Vec<Felt252>,
+//     pub data: Vec<Felt252>,
```
Comment (odesenfans) [inline src/execution/syscalls.rs:114]: Why did you change commented code?

**Review by odesenfans — CHANGES_REQUESTED** [review]

File: Cargo.lock [hunk 3]
```diff
@@ -965,7 +965,7 @@ dependencies = [
 [[package]]
 name = "cairo-vm"
 version = "1.0.0-rc1"
-source = "git+https://github.com/lambdaclass/cairo-vm#ec00e31ec5ed412d520957edef8b3441db82d133"
+source = "git+https://github.com/lambdaclass/cairo-vm#c692b7594a74ee3245cf7affc7c8cfc296b0398c"
```
[the remark this answers is not in the document]
Comment (maciejka) [reply Cargo.lock:968]: He have:
```cairo-vm = { git = "https://github.com/lambdaclass/cairo-vm", features = ["extensive_hints", "cairo-1-hints"] }```
in Cargo.toml which does not specify actual commit, and my local version did not have any of your prs recently merged into the vm, so I had to do the update. If there is a better way...
Staying on top of vm changes is not a bad thing at the current stage of the project. imho.

File: src/execution/syscall_utils.rs [hunk 4]
```diff
@@ -114,6 +114,14 @@ pub fn read_calldata(vm: &VirtualMachine, ptr: &mut Relocatable) -> SyscallResul
     read_felt_array::<SyscallExecutionError>(vm, ptr)
 }
 
+pub fn ignore_felt_array<TErr>(_vm: &VirtualMachine, ptr: &mut Relocatable) -> Result<(), TErr>
```
[the remark this answers is not in the document]
Comment (maciejka) [reply src/execution/syscall_utils.rs]: You are right.

File: src/execution/syscall_utils.rs [hunk 5]
```diff
@@ -114,6 +114,14 @@ pub fn read_calldata(vm: &VirtualMachine, ptr: &mut Relocatable) -> SyscallResul
...
+where
+    TErr: From<VirtualMachineError> + From<MemoryError> + From<MathError>,
+{
+    *ptr = (*ptr + 2)?;
```
[the remark this answers is not in the document]
Comment (maciejka) [reply src/execution/syscall_utils.rs:119]: Array `ptr` is pointing to is represented by two pointers that is why `+2`. Handling `emit_event` syscall does not need reading content of the array that is why function is called ignore. I prefer to keep the function as it kind of documents what is going on.

[the remark this answers is not in the document]
Comment (odesenfans) [reply src/execution/syscall_utils.rs:119]: Okay, but then this should be documented in some way to explain what the array contains and why 2 is the right value.

File: src/execution/syscalls.rs (hunk 1 above)
Comment (maciejka) [reply src/execution/syscalls.rs:124]: `syscall_handler.py`:

```
    def emit_event(self, remaining_gas: int, request: CairoStructProxy) -> SyscallFullResponse:
        keys = self._get_felt_range(start_addr=request.keys_start, end_addr=request.keys_end)
        data = self._get_felt_range(start_addr=request.data_start, end_addr=request.data_end)
        self._emit_event(keys=keys, data=data)

        response_header = self.structs.ResponseHeader(gas=remaining_gas, failure_flag=0)
        return response_header, tuple()

```
and:
```
    def _emit_event(self, keys: List[int], data: List[int]):
        return
```

File: src/execution/syscalls.rs (hunk 2 above)
Comment (maciejka) [reply src/execution/syscalls.rs:114]: It was previous version of `EmitEventRequest`. Removed.

File: src/execution/syscalls.rs (hunk 1 above)
Comment (odesenfans) [reply src/execution/syscalls.rs:117]: That's what I thought. But then, why do you not get the `keys` and `data` arrays?

Comment (maciejka) [reply src/execution/syscalls.rs:117]: Because emit_even in replay phase does not need them. See empty `_emit_event` above.

File: src/execution/syscall_utils.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (maciejka) [reply src/execution/syscall_utils.rs:115]: I moved the function a few lines up, to put it just below `read_felt_array` which actually reads the array data. I think that in its context `ignore_felt_array` is self explanatory.

[the remark this answers is not in the document]
Comment (odesenfans) [reply src/execution/syscall_utils.rs:115]: For me the +2 is not self explanatory.

File: src/execution/syscalls.rs (hunk 1 above)
Comment (maciejka) [reply src/execution/syscalls.rs:117]: TODO for what?

File: src/execution/syscall_utils.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (maciejka) [reply src/execution/syscall_utils.rs:116]: added comment

File: src/execution/syscalls.rs (hunk 1 above)
Comment (odesenfans) [reply src/execution/syscalls.rs:117]: We will never need to implement the functionality of the syscall directly?

Comment (maciejka) [reply src/execution/syscalls.rs:117]: No. Here we are just replying return values in order to get contract logic behave as expected. Syscalls are reexecuted in cairo in a later stage:
image

**Review by odesenfans — APPROVED** [review]
